Add test cases for all LogMsg routes in logs.js

omarraf opened the pull request:
# Description

Adds test coverage for all routes in /src/server/routes/logs.js

Tested Routes include:
- POST /api/logs/info
- POST /api/logs/warn
- POST /api/logs/error
- GET /api/logs/logsmsg/getLogsByDateRangeAndType

All tests pass locally on Docker: 

<img width="556" alt="image" src="https://github.com/user-attachments/assets/5d48fa93-6f00-48ee-815e-9a5dcff161d5" />


Fixes #1410

## Type of change

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

omarraf committed:
Add test cases for all LogMsg routes in logs.js

omarraf committed:
Merge branch 'OpenEnergyDashboard:development' into feature/add-log-route-tests

huss reviewed (commented):
Thanks to @omarraf for addressing this issue. I have not been able to fully review the proposed changes but do have one thought. What do you think about having tests where log msgs are inserted and then requests are made for different types of logs and different time ranges to see if the proper msgs are returned. This would require tracking the inserted msgs, filtering them and comparing to what is returned. OED does some tests of this type in other places. I welcome your thoughts and any questions.

omarraf commented:
Thanks for getting back to me @huss 

> What do you think about having tests where log msgs are inserted and then requests are made for different types of logs and different time ranges to see if the proper msgs are returned

I agree this would provide much better test coverage. The current tests only verify the API surface but not the underlying functionality. 

> any questions

I was wondering if I should insert the test log data directly using the LogMsg model, or create it through the API endpoints first?

huss commented:
Inserting via the API/route will test both the route and model. Since the tests rarely fail, testing both at the same time seems easier and what OED often does. In a perfect world the insert and retrieve would be separate but it seems ok to do both together. If a test fails then someone can determine what part is the issue. Does that make sense and seem ok?
Summary: Use the API/route to insert the log msgs and to retrieve in the tests.

omarraf commented:
Yes that makes sense. Thanks for the clarification!

omarraf committed:
Merge branch 'OpenEnergyDashboard:development' into feature/add-log-route-tests

omarraf committed:
enhance route tests with insert and verify functionality

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+const moment = require('moment');
+
+mocha.describe('Log Routes', () => {
+   let token;
```
I know the test files are inconsistent but OED prefers tab indenting. VSC can do with "Convert Indentation to Tabs".

huss replied:
I'm still seeing spaces. If you want me to fix this then I can or you are welcome to do it.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+const { chai, mocha, expect, app, testUser } = require('../common');
```
The MPL license comment in other files needs to be added.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+       token = res.body.token;
+   });
+
+   // Original status code tests
```
I think this comment can go.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+           const response = await chai.request(app)
+               .post('/api/logs/info')
+               .set('token', token)
+               .send({ message: 'Valid log message' });
```
Maybe for consistency this could be 'Valid info message'. Also similar for the mocha.it above for this test. Similarly for invalid info below.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+           const response = await chai.request(app)
+               .get('/api/logs/logsmsg/getLogsByDateRangeAndType')
+               .set('token', token)
+               .query({ timeInterval: '2023-01-01T00:00:00Z', logTypes: 'INFO', logLimit: '10' });
```
While the timeInterval values seem to work, they are normally a start and end value. See createTimeString in src/server/util/readingsUtils.js. I suspect it is defaulting to half unbounded (end time is +infinity) but have not checked. Actual values for both might be better for testing (unless doing the unbounded test).

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+               });
+
+           expect(infoResponse.status).to.equal(200);
+           const infoLogs = infoResponse.body.filter(log => 
```
I'm unclear what the filtering does. The only returned log should be the info one. I think testing without filtering would be a stronger test and still work. This applies to more tests below.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+           expect(warnLogs).to.have.lengthOf(1);
+           expect(warnLogs[0].logType).to.equal('WARN');
+           expect(warnLogs[0].logMessage).to.equal(warnMessage);
+       });
```
Is there a reason you don't check for error logmsgs?

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+           // Should contain INFO and WARN logs, but not ERROR
+           expect(testLogs).to.have.lengthOf(2);
+           const logTypes = testLogs.map(log => log.logType).sort();
+           expect(logTypes).to.deep.equal(['INFO', 'WARN']);
```
Also testing the actual log msg would make this a stronger test.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+           expect(response.body).to.be.an('array');
+           
+           // Should not return more than the limit
+           expect(response.body.length).to.be.at.most(2);
```
I think in this case it should be exactly 2 since you inserted 5. If it was less than 2 then it made a mistake.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+
+       mocha.it('should return empty array when no logs match filters', async () => {
+           // Query for logs with a very specific type that doesn't exist
+           const response = await chai.request(app)
```
I'm wondering if this would be a stronger test if other log msgs were inserted so some existed that are not returned.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,302 @@
+                   logTypes: 'INFO',
+                   logLimit: '100'
+               });
+           
```
There are extra spaces on this blank line. VSC "Format Document" would fix any of these and other formatting issues (if exist).

huss reviewed (changes requested):
Thanks to @omarraf for this work. Testing found that overall it works well. I have made a few comments to consider. Please let me know if I can help in any way.

huss commented:
@omarraf It has been a couple of weeks since I posted some comments. Thus, I wanted to touch base to see if you needed any help or were continuing to work on this. We would appreciate an update on the status and appreciate your efforts on this work.

omarraf commented:
Hey @huss, thanks for checking in. I'm currently working on improving the code per your comments, I'll have an updated version with all of the needed changes by this week.

huss commented:
@omarraf Thanks for the update. I'm not trying to rush you but wanted to see what is up. I look forward to seeing your ideas/changes.

omarraf committed:
add MPL license

omarraf committed:
specify log type in message

omarraf committed:
fix: time interval, remove weak filtering, add error tests

omarraf committed:
format document and add improve no logs match filters test

omarraf committed:
add proper time intervals

omarraf committed:
convert indentation to tabs

omarraf commented:
@huss I have addressed all of the comments in my latest commits, please let me know if I missed something or any suggestions to make the tests stronger. Thanks!

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,359 @@
+
+            expect(infoResponse.status).to.equal(200);
+
+            // Should only contain INFO logs
```
I appreciate your change. I think this code might be simpler and I found it worked the same. This is what I was thinking to remove the filters.
```
			expect(infoResponse.status).to.equal(200);
			// Should only contain one INFO log
			const infoBody = infoResponse.body;
			expect(infoBody.length).to.equal(1);
			const testLog = infoBody[0];
			expect(testLog.logType).to.equal('INFO');
			expect(testLog.logMessage).to.equal(infoMessage);
```
It replaces everything past the info query. What do you think?

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,359 @@
+            expect(warnLog.logType).to.equal('WARN');
+            expect(warnLog.logMessage).to.equal(warnMessage);
+
+            expect(errorLog).to.be.undefined; // Should not be returned
```
Small point but OED prefers comments above the code and not at the end of the line.

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,359 @@
+            const response = await chai.request(app)
+                .get('/api/logs/logsmsg/getLogsByDateRangeAndType')
+                .set('token', token)
+                .query({ timeInterval: '2023-01-01T00:00:00Z_2023-12-31T23:59:59Z', logTypes: 'INFO', logLimit: '10' });
```
This works so thanks. I'm just noting that it depends on the format used by the toString in src/common/TimeInterval.js. It is a little more code but if a TimeInterval instance was created and then toString from that class was used then the code would always use whatever format it thinks is best. I suspect changes might be unlikely but it is slightly better practice. Note createTimeString in src/server/util/readingsUtils.js does this.

huss reviewed (changes requested):
Thanks to @omarraf for addressing my previous comments. Overall it looks good. I've made a few new comment to consider. Please let me know if you have questions or would like help.

omarraf committed:
add simpler test for info log query

omarraf committed:
clean up comments

huss commented on src/server/test/routes/logsRouteTests.js:
```diff
@@ -0,0 +1,356 @@
+					logLimit: '100'
+				});
+
+			expect(warnResponse.status).to.equal(200);
```
I think warn and error should use the same, new logic for info. Sorry I was not clear about that.

omarraf replied:
No worries! I'll update it right away, thank you for all your helpful comments.

huss reviewed (changes requested):
@omarraf Thanks for addressing my previous comments. This is looking good. I had one comment to consider and then this should be ready to merge. Please let me know if you need anything.

omarraf committed:
improve warn and error log tests

omarraf commented:
I appreciate the feedback, I addressed the latest comments with my last commit. Let me know if there's anything else I can do.

huss reviewed (approved):
Thanks to @omarraf for the update and seeing this through to completion. Review and testing found it works as expected. Congratulations on having your work accepted by OED.

huss commented:
I'm noting that the automated tests are failing for 3 pipeline test. This seems to be a general problem that I am looking into. I'm going to hold off merging this work to see if I can resolve first but I'm certain it has nothing to do with this work.

huss commented:
> I'm noting that the automated tests are failing for 3 pipeline test. This seems to be a general problem that I am looking into. I'm going to hold off merging this work to see if I can resolve first but I'm certain it has nothing to do with this work.

I don't want to hold up PRs for too long. Since I have not found the time to investigate this yet and I'm certain that there are no issues with this PR, I'm going to merge it with the failed tests.
